schemas/v1/openapi.yaml

```diff
@@ -0,0 +1,241 @@
+        '400':
+          description: The given request is not valid.
+          content:
+            text/plain: { }
+        '404':
+          description: No results were found for any of the given multihashs.
+        '500':
+          description: Failure occurred while processing the request.
+          content:
+            text/plain: { }
+components:
+  schemas:
+    FindResponse:
+      type: array
+      items:
+        $ref: '#/components/schemas/MultihashResults'
+    MultihashResults:
+      type: array
+      items:
+        type: object
+        properties:
+          Multihash:
+            type: string
+            description: base64 string representation of the queried multihash
+          ProviderResults:
+            type: array
+            items:
+              $ref: '#/components/schemas/ProviderRecord'
+    ProviderRecord:
+      type: object
+      properties:
+        ContextID:
+          type: string
+          description: base64 encoded context ID.
+        Metadata:
+          type: string
+          description: base64 encoded metadata.
+        Provider:
+          type: object
+          properties:
+            ID:
+              type: string
+              description: encoded libp2p peer ID of the provider.
+            Addrs:
+              type: array
+              items:
+                type: string
+                description: string representation of the provider multiaddr.
+    FindRequest:
+      type: object
+      properties:
+        Multihashes:
+          type: array
+          description: the list of multihashes for which to find providers.
+          items:
+            type: string
+            description: base64 string representation of the multihash.
+  examples:
+    NDJsonProviderRecords:
+      summary: Streaming provider records response
+      value: |
+        {"ContextID":"aXBmcy1kaHQtY2FzY2FkZQ==","Metadata":"gBI=","Provider":{"ID":"12D3KooWHVXoJnv2ifmr9K6LWwJPXxkfvzZRHzjiTZMvybeTnwPy","Addrs":["/ip4/145.40.89.101/tcp/4001","/ip4/145.40.89.101/tcp/4002/ws","/ip4/145.40.89.101/udp/4001/quic","/ip6/2604:1380:45f1:d800::1/tcp/4001","/ip6/2604:1380:45f1:d800::1/tcp/4002/ws","/ip6/2604:1380:45f1:d800::1/udp/4001/quic"]}}
+        
+        {"ContextID":"aXBmcy1kaHQtY2FzY2FkZQ==","Metadata":"gBI=","Provider":{"ID":"12D3KooWDpp7U7W9Q8feMZPPEpPP5FKXTUakLgnVLbavfjb9mzrT","Addrs":["/ip4/147.75.80.75/tcp/4001","/ip4/147.75.80.75/tcp/4002/ws","/ip4/147.75.80.75/udp/4001/quic","/ip6/2604:1380:4601:f600::5/tcp/4001","/ip6/2604:1380:4601:f600::5/tcp/4002/ws","/ip6/2604:1380:4601:f600::5/udp/4001/quic"]}}
+        
+        {"ContextID":"aXBmcy1kaHQtY2FzY2FkZQ==","Metadata":"gBI=","Provider":{"ID":"12D3KooWCrBiagtZMzpZePCr1tfBbrZTh4BRQf7JurRqNMRi8YHF","Addrs":["/ip4/147.75.87.65/tcp/4001","/ip4/147.75.87.65/tcp/4002/ws","/ip4/147.75.87.65/udp/4001/quic","/ip6/2604:1380:4601:f600::1/tcp/4001","/ip6/2604:1380:4601:f600::1/tcp/4002/ws","/ip6/2604:1380:4601:f600::1/udp/4001/quic"]}}
```

Comment:
A possible optimization would be to only have the provider ID, and not include the provider addresses, if the provider had already been returned earlier in the stream.

Reply:
I am worried that it could make the client side more complex, specially in a case where results are distributed across threads or even retriever instances.

Happy to revisit if the existing improvements prove insufficient.